packages/ui/react-ui-table/src/model/table-presentation.ts

```diff
@@ -0,0 +1,202 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { signal } from '@preact/signals-core';
+
+import { FormatEnum, getValue } from '@dxos/echo-schema';
+import { cellClassesForFieldType, cellClassesForRowSelection, formatForDisplay } from '@dxos/react-ui-form';
+import {
+  type DxGridPlane,
+  type DxGridPlaneCells,
+  toPlaneCellIndex,
+  type DxGridPlaneRange,
+  type DxGridCellValue,
+} from '@dxos/react-ui-grid';
+import { mx } from '@dxos/react-ui-theme';
+import { VIEW_FIELD_LIMIT, type FieldType } from '@dxos/schema';
+
+import { type BaseTableRow, type TableModel } from './table-model';
+import { tableButtons } from '../util';
+import { tableControls } from '../util/table-controls';
+
+export class TablePresentation<T extends BaseTableRow = { id: string }> {
+  constructor(
+    private readonly model: TableModel<T>,
+    private readonly _visibleRange = signal<DxGridPlaneRange>({
+      start: { row: 0, col: 0 },
+      end: { row: 0, col: 0 },
+    }),
+  ) {}
+
+  public getCells = (range: DxGridPlaneRange, plane: DxGridPlane): DxGridPlaneCells => {
+    switch (plane) {
+      case 'grid': {
+        this._visibleRange.value = range;
+        return this.getMainGridCells(range);
+      }
+      case 'frozenRowsStart': {
+        return this.getHeaderCells(range);
+      }
+      case 'frozenColsStart': {
+        return this.getSelectionColumnCells(range);
+      }
+      case 'frozenColsEnd': {
+        return this.getActionColumnCells(range);
+      }
+      case 'fixedStartStart': {
+        return this.getSelectAllCell();
+      }
+      case 'fixedStartEnd': {
+        return this.getNewColumnCell();
+      }
+      default: {
+        return {};
```

Comment:
is this a valid result?

Reply:
Yes. It's typed.

Reply:
We supply empty record for all grid planes we aren't interested in filling.